Approved/Initializers.md

```diff
@@ -0,0 +1,175 @@
+Since it is a common scenario to initialize qubits to a certain state, this proposal includes the means to construct initializers on the fly that do just that. There are two distinct cases that are common. 
+1. One is that an operation should be invoked on the allocated qubits after initialization, followed by further computations before the qubits are measured and released. In this case, it is not necessary or desirable that the operation performed upon initialization is un-computed upon release. 
+2. In the second case, the computations following the initialization lead to the qubits being in the same state as they were initialized. In this case, the adjoint of the operation used for initialization needs to be applied upon releasing the qubits. 
+
+The proposal is to syntactically distinguish the two cases. The table below contains examples for both cases. In the first case, the initializer consists of the keyword `init`, followed by the initializer argument tuple, the keyword `then` and an operation-valued expression. The initializer argument tuple may be omitted if the required argument is of type `Unit`. The second case follows the same pattern with the exception that instead of the keyword `then`, then keyword `within` should be used, and the operation-valued expression needs to support the `Adjoint` functor.   
+In both cases, the elevated operation has to return `Unit`.
+The allocated value matches the argument of the elevated operation and will be bound to the symbol(s) specified on the left hand side of the allocation statement.  
+
+### *Initializer expressions and nesting of initializers*:    
+An initializer expressions consists either of a call to an initializer, or of a tuple of multiple calls. ...
+
+### Overview over different kinds of initializer expressions
+
+Initializers for built-in types:
+| Built-in Initializer | Description | Allocated Type | Initializer Argument Type | Initializer Expression | 
+| --- | --- | --- | --- | --- |
+| Qubit | allocates a single qubit | `Qubit` | `Unit` | `Qubit()` |
+| Qubits | allocates a 1D array of qubits | `Qubit[]` | `Int` | `Qubits(4)` |
+| Qubits | allocates a [2D array](https://github.com/microsoft/qsharp-language/issues/39) of qubits | `Qubit[,]` | `(Int, Int)` | `Qubits(2,4)` |
+
+Generated default initializers for custom types:
+| Allocated Type | Type of Contained Items | Initializer Argument Type | Initializer Expression |
+| --- | --- | --- | --- |
+| BigEndian | `Qubit[]` | `Int` | `BigEndian(4)` |
+| QubitArray | `(Int, Qubit[])` | `(Int, Int)` | `QubitArray(4, 4)` |
+
+Initializers built by elevating operations:
+| Operation Valued Expression | Allocated Type | Initializer Argument Type | Initializer Expression | Uncompute upon Release |
+| --- | --- | --- | --- | --- |
+| `H` | `Qubit` | `Unit` | `init within H` | Yes |
+| `Reset` | `Qubit` | `Int` | `init then Reset` | No |
+| `ApplyToEachA(H, _)` | `Qubit[]` | `Int` | `init(3) within ApplyToEachA(H, _)` | Yes |
```

Comment:
If it's in scope, it may be helpful to use `ApplyXorInPlace` as an example, since that demonstrates how `init`/`then` can be used to really help out with preparing structured data. In particular, I think that builds well on the example at line 68.